torch/distributed/distributed_c10d.py

```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            applicable only if the environment variable ``NCCL_BLOCKING_WAIT``
+            or ``NCCL_ASYNC_ERROR_HANDLING`` is set to 1. When
+            ``NCCL_BLOCKING_WAIT`` is set, this is the duration for which the
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
+            this is the duration after which collectives will be aborted
+            asynchronously and the process will crash. ``NCCL_BLOCKING_WAIT``
+            will provide errors to the user which can be caught and handled,
+            but due to its blocking nature, it has a performance overhead. On
+            the other hand, ``NCCL_ASYNC_ERROR_HANDLING`` has very little
+            performance overhead, but crashes the process on errors. This is
+            done since CUDA execution is async and it is no longer safe to
+            continue executing user code since failed async NCCL operations
+            might result in subsequent CUDA operations running on corrupted
+            data. Only one of these two environment variables should be set.
+        backend (str or Backend, optional): The backend to use. Depending on
+            build-time configurations, valid values are ``gloo`` and ``nccl``.
+            By default uses the same backend as the global group. This field
+            should be given as a lowercase string (e.g., ``"gloo"``), which can
+            also be accessed via :class:`Backend` attributes (e.g.,
+            ``Backend.GLOO``). If ``None`` is passed in, the backend
+            corresponding to the default process group will be used. Default is
+            ``None``.
+        pg_options (ProcessGroupOptions, optional): process group options
+            specifying what additional options need to be passed in during
+            the construction of specific process groups. i.e. for the ``nccl``
+            backend, is_high_priority_stream can be specified so that process
+            group can pick up high priority cuda streams.
+
+    Returns:
+        The subgroup containing the current rank, and all the subgroups used for cleanup.
+    """
+    subgroups = []
+    cur_subgroup = None
+    if ranks_per_subgroup_list:
+        for ranks in ranks_per_subgroup_list:
+            subgroup = new_group(
+                ranks=ranks, timeout=timeout, backend=backend, pg_options=pg_options
+            )
+            subgroups.append(subgroup)
+
+            rank = get_rank()
+            if rank in ranks:
+                cur_subgroup = subgroup
+                logger.info("Rank {} is assigned to subgroup {}".format(rank, ranks))
+        return cur_subgroup, subgroups
+
+    # By default, create one subgroup per machine.
+    # Assume that each machine has the same number of devices.
+    num_devices_per_machine = (
+        torch.cuda.device_count()
+        if torch.cuda.is_available()
+        else (os.cpu_count() or 1)
```

Comment:
I'm actually not sure about what the expected behavior should be for CPU-only use cases. I don't think I've seen a use case where one rank is spawned for each CPU core, which is what this code seems to do/ Is there such a use case? If not, should we always just use `1` if cuda is not available?

Reply:
If a machine has multiple CPUs and no GPU, then each subgroup can have multiple ranks where each rank corresponds to a CPU. I think this can avoid surprising users and keep the subgroup creation behavior consistent in GPU and CPU-only scenarios.

I haven't seen any CPU-only use case yet, so we cannot easily know what will be expected behavior. I will prefer keeping the code this way.

Added a warning to declare that this API is experimental and subject to change, so we can make changes in the future.